src/layer/x86/layernorm_x86.cpp

```diff
@@ -0,0 +1,483 @@
+//
+// Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
+//
+// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except
+// in compliance with the License. You may obtain a copy of the License at
+//
+// https://opensource.org/licenses/BSD-3-Clause
+//
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#include "layernorm_x86.h"
+#include "x86_usability.h"
+#include <math.h>
+#include <cpu.h>
+
+#if __SSE2__
+#include <emmintrin.h>
+#if __AVX__
+#include <immintrin.h>
+#endif // __AVX__
+#endif // __SSE2__
+
+static NCNN_FORCEINLINE void fast_mean(float* ptr, float* mean, int elempack, int elemcount, int size)
+{
+    int i = 0;
+    float sum = 0.0f;
+#if __SSE2__
+#if __AVX__
+#if __AVX512F__
+    if (elempack == 16 || elempack == 1)
+    {
+        __m512 _sum = _mm512_setzero_ps();
+        for (; i + 16 <= size; i += 16, ptr += 16)
+        {
+            __m512 _cur = _mm512_loadu_ps(ptr);
+            _sum = _mm512_add_ps(_sum, _cur);
+        }
+        if (elempack == 16)
+        {
+            __m512 _elemcount = _mm512_set1_ps(float(elemcount));
+            __m512 _mean = _mm512_div_ps(_sum, _elemcount);
+            _mm512_storeu_ps(mean, _mean);
+        }
+        else
+        {
+            sum += _mm512_reduce_add_ps(_sum);
+        }
+    }
+#endif // __AVX512F__
+    if (elempack == 8 || elempack == 1)
+    {
+        __m256 _sum = _mm256_setzero_ps();
+        for (; i + 8 <= size; i += 8, ptr += 8)
+        {
+            __m256 _cur = _mm256_loadu_ps(ptr);
+            _sum = _mm256_add_ps(_sum, _cur);
+        }
+        if (elempack == 8)
+        {
+            __m256 _elemcount = _mm256_set1_ps(float(elemcount));
+            __m256 _mean = _mm256_div_ps(_sum, _elemcount);
+            _mm256_storeu_ps(mean, _mean);
+        }
+        else
+        {
+            sum += _mm256_reduce_add_ps(_sum);
+        }
+    }
+#endif // __AVX__
+    if (elempack == 4 || elempack == 1)
+    {
+        __m128 _sum = _mm_setzero_ps();
+        for (; i + 4 <= size; i += 4, ptr += 4)
+        {
+            __m128 _cur = _mm_loadu_ps(ptr);
+            _sum = _mm_add_ps(_sum, _cur);
+        }
+        if (elempack == 4)
+        {
+            __m128 _elemcount = _mm_set1_ps(float(elemcount));
+            __m128 _mean = _mm_div_ps(_sum, _elemcount);
+            _mm_storeu_ps(mean, _mean);
+        }
+        else
+        {
+            sum += _mm_reduce_add_ps(_sum);
+        }
+    }
+#endif // __SSE2__
+    if (elempack == 1)
+    {
+        for (; i < size; ++i, ++ptr)
+        {
+            sum += *ptr;
+        }
+        *mean = sum / elemcount;
+    }
+}
```

Comment:
This may not be what I expected :P

we can make the main loop benefit from any simd isa available, no matter what the elempack is, but handle the elempack thing in post processing block.

sample code

```cpp
    int i = 0;
#if __SSE2__
#if __AVX__
#if __AVX512F__
    __m512 _sum_avx512 = _mm512_setzero_ps();
    for // step 16
    {
        // add to _sum_avx512
    }
#endif // __AVX512F__
    __m256 _sum_avx = _mm256_setzero_ps();
    for // step 8
    {
        // add to _sum_avx
    }
#endif // __AVX__
    __m128 _sum = _mm_setzero_ps();
    for // step 4
    {
        // add to _sum
    }
#endif // __SSE2__
    float sum = 0.0f;
    for // step 1
    {
        // add to sum
    }

    // ok, now we have _sum_avx512 _sum_avx _sum and sum
    // handle for different elempack
#if __SSE2__
#if __AVX__
#if __AVX512F__
    if (elempack == 16)
    {
        // write _sum_avx512 to mean
    }
#endif // __AVX512F__
    if (elempack == 8)
    {
#if __AVX512F__
        // split _sum_avx512 into two parts and add to _sum_avx
#endif // __AVX512F__
        // write _sum_avx to mean
    }
#endif // __AVX__
    if (elempack == 4)
    {
#if __AVX__
#if __AVX512F__
        // split _sum_avx512 into two parts and add to _sum_avx
#endif // __AVX512F__
        // split _sum_avx into two parts and add to _sum
#endif // __AVX__
        // write _sum to mean
    }
#endif // __SSE2__
    if (elempack == 1)
    {
#if __SSE2__
#if __AVX__
#if __AVX512F__
        // horizontal _sum_avx512 and add to sum
#endif // __AVX512F__
        // horizontal _sum_avx and add to sum
#endif // __AVX__
        // horizontal _sum and add to sum
#endif // __SSE2__
        // write sum to mean
    }
```

Reply:
I've considered such plans. But writing that way makes codes as complex as the original draft (where I did not combine many branches into one).
I think as long as the ISA supports, the `elempack` would be as broad as possible. (Is that correct?) If that's true, in the first part of your sample codes where it sums no matter what the `elempack` is, packed data would only be summed in certain one for loop block and unpacked data would be summed horizontally across different blocks. I think our codes do the same thing :P. My fallback condition `(... || elempack == 1)` would ensure unpacked data also benefits from SIMD operations.

Reply:
consider elempack=4 case, it may be a tensor with channel 124, which can not be perfectly divided by 16.
In the current implementation, it will fallback to __m128 block, even avx/avx512 is supported.

Reply:
I see that. Newer commits cancel the relation between SIMD instruction set and `elempack`.